File: src/main/java/com/alexandre/splittrip/model/Expense.java
package com.alexandre.splittrip.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
public class Expense {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String description;

    private BigDecimal totalAmount;

    private LocalDateTime expenseDate = LocalDateTime.now();

    @JsonIgnore
    @ManyToOne
    @JoinColumn(name = "trip_id", nullable = false)
    private Trip trip;

    @ManyToOne
    @JoinColumn(name = "paid_by_member_id", nullable = false)
    private Member paidBy;

    @OneToMany(mappedBy = "expense", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ExpenseSplit> splits;

    public Expense() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public BigDecimal getTotalAmount() { return totalAmount; }
    public void setTotalAmount(BigDecimal totalAmount) { this.totalAmount = totalAmount; }
    public LocalDateTime getExpenseDate() { return expenseDate; }
    public void setExpenseDate(LocalDateTime expenseDate) { this.expenseDate = expenseDate; }
    public Trip getTrip() { return trip; }
    public void setTrip(Trip trip) { this.trip = trip; }
    public Member getPaidBy() { return paidBy; }
    public void setPaidBy(Member paidBy) { this.paidBy = paidBy; }
    public List<ExpenseSplit> getSplits() {return splits;}
    public void setSplits(List<ExpenseSplit> splits) {this.splits = splits;}
}

File: mobile/lib/models/expense.dart
import 'member.dart';
import 'expense_split.dart';

class Expense {
  final int id;
  final String description;
  final double totalAmount;
  final Member paidBy;
  final List<ExpenseSplit> splits;

  Expense({
    required this.id,
    required this.description,
    required this.totalAmount,
    required this.paidBy,
    this.splits = const [],
  });

  factory Expense.fromJson(Map<String, dynamic> json) {
    return Expense(
      id: json['id'] as int,
      description: json['description'] as String,
      totalAmount: (json['totalAmount'] as num).toDouble(),
      paidBy: Member.fromJson(json['paidBy'] as Map<String, dynamic>),
      splits: (json['splits'] as List<dynamic>?)
              ?.map((e) => ExpenseSplit.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}